Add docs about configuring SSL for the Logstash output (#1807)
elastic/observability-docs: docs/en/ingest-management/security/logstash-certificates.asciidoc

{agent} clients.

[discrete]
[[secure-logstash-prereqs]]
== Prerequisites

* Make sure your https://www.elastic.co/subscriptions[subscription level]
supports output to {ls}.

* On Windows, add port 8220 for {fleet-server} and 5044 for Logstash to the
inbound port rules in Windows Advanced Firewall.

* If you are connecting to a self-managed {es} cluster, you need the CA
certificate that was used to sign the certificates for the HTTP layer of {es}
cluster. For more information, refer to the
{ref}/configuring-stack-security.html[{es} security docs].

[discrete]
[[generate-logstash-certs]]
== Generate custom certificates and private keys

You can use whatever process you typically use to generate PEM-formatted
certificates. The examples shown here use the `certutil` tool provided by {es}.

TIP: The `certutil` tool is not available on {ecloud}, but you can still use it
to generate certificates for {agent} to {ls} connections. Just
https://www.elastic.co/downloads/elasticsearch[download an {es} package],
extract it to a local directory, and run the `elasticsearch-certutil` command.
There's no need to start {es}!

> 
> //REVIEWERS: I did not set up /etc/hosts. Please help me refine the steps here.
> //for example, do I really need to set --ip and --dns? Or is there a better way?
> //Because I'm running all on the same machine (and using docker for the agent
> //deployment, I might be missing something.)

Author: ```suggestion
```

[[secure-logstash-connections]]
= Configure SSL/TLS for the {ls} output

beta[]

To send data from {agent} to {ls} securely, you need to configure Transport
Layer Security (TLS). Using TLS ensures that your {agent}s send encrypted data
to trusted {ls} servers, and that your {ls} servers receive data from trusted
{agent} clients.

[discrete]
[[secure-logstash-prereqs]]
== Prerequisites

//REVIEWERS: Is there a specific subscription level required for this? The issue
//mentions platinum, but as a general rule we point to the subsciptions page.

> //It's not clear to me which part of this feature requires plat.

Author: OK. I think a link to the subscriptions page is the right thing to do here. I've already documented that per-policy outputs require a specific subscription level, too, without mentioning platinum specifically.

> 
> //REVIEWERS: Is there a specific subscription level required for this? The issue
> //mentions platinum, but as a general rule we point to the subsciptions page.
> //It's not clear to me which part of this feature requires plat.

Author: ```suggestion
```

* Make sure your https://www.elastic.co/subscriptions[subscription level]
supports output to {ls}.

* On Windows, add port 8220 for {fleet-server} and 5044 for Logstash to the
inbound port rules in Windows Advanced Firewall.

* If you are connecting to a self-managed {es} cluster, you need the CA
certificate that was used to sign the certificates for the HTTP layer of {es}
cluster. For more information, refer to the
{ref}/configuring-stack-security.html[{es} security docs].

> 
> //REVIEWERS: I'm not going to explain the steps for securing the LS to ES
> //connection here because IMO that's beyond the scope of this doc and may
> //confused cloud users who don't need to know this.

Author: ```suggestion
```

[discrete]
[[generate-logstash-certs]]
== Generate custom certificates and private keys

You can use whatever process you typically use to generate PEM-formatted
certificates. The examples shown here use the `certutil` tool provided by {es}.

> 
> //REVIEWERS: The following worked for me, and I saw people asking on discuss. Is
> //there any reason we should not recommend this to cloud users? 

Author: ```suggestion
```

TIP: The `certutil` tool is not available on {ecloud}, but you can still use it
to generate certificates for {agent} to {ls} connections. Just
https://www.elastic.co/downloads/elasticsearch[download an {es} package],
extract it to a local directory, and run the `elasticsearch-certutil` command.
There's no need to start {es}!

//REVIEWERS: I did not set up /etc/hosts. Please help me refine the steps here.
//for example, do I really need to set --ip and --dns? Or is there a better way?
//Because I'm running all on the same machine (and using docker for the agent

> //deployment, I might be missing something.)

Reviewer: I think if you're not setting up `/etc/hosts` then the `--dns` option is not necessary, but I may be wrong. The `--ip` is likely necessary if you're using Docker for the setup, since it'll need to be set for the Docker gateway. e.g. `192.168.65.2`. 

These options specify the valid host names and IP addresses for which the certificates generated will be considered valid.

Author: I'm going to leave this as-is. I think this is OK for beta at least.

. Generate a certificate authority (CA). Skip this step if you want to use an
existing CA.
+
--
[source,shell]
----
./bin/elasticsearch-certutil ca --pem
----

This command creates a zip file that contains the CA certificate and key you'll
use to sign the certificates. Extract the zip file:

image::images/ca-certs.png[Screen capture of a folder called ca that contains two files: ca.crt and ca.key]
--

. Generate a client SSL certificate signed by your CA. For example:
+
--
[source,shell]
----
./bin/elasticsearch-certutil cert \
  --name client \
  --ca-cert /path/to/ca/ca.crt \
  --ca-key /path/to/ca/ca.key \
  --dns your.host.name.here \
  --ip 192.0.2.1 \
  --pem
----

> Where `dns` and `ip` specify the name and IP address of the client.

Author: ```suggestion
Where `dns` and `ip` specify the valid host names and IP addresses where the generated certificates are valid.
```

Extract the zip file:

image::images/client-certs.png[Screen capture of a folder called client that contains two files: client.crt and client.key]
--

. Generate a {ls} SSL certificate signed by your CA. For example:
+
--
[source,shell]
----
./bin/elasticsearch-certutil cert \
  --name logstash \
  --ca-cert /path/to/ca/ca.crt \
  --ca-key /path/to/ca/ca.key \
  --dns your.host.name.here \
  --ip 192.0.2.1 \
  --pem
----

> Where `dns` and `ip` specify the name and IP address of the {ls} server.

Author: ```suggestion
```

Extract the zip file:

image::images/logstash-certs.png[Screen capture of a folder called logstash that contains two files: logstash.crt and logstash.key]
--

. Convert the {ls} key to pkcs8. For example, on Linux run:
+
[source,shell]
----
openssl pkcs8 -inform PEM -in logstash.key -topk8 -nocrypt -outform PEM -out logstash.pkcs8.key
----

//REVIEWERS: The GitHub issue says that Linux must be used to convert
//the key. Is this true? Is what I have here ^^ accurate and good enough for Win

> //users?

Reviewer: I'm not sure about this. I was able to convert the key to `pkcs8` on my Mac without issue. I believe the `openssl` executable should still work as expected here on Windows based on a quick look around Google.

> 
> //REVIEWERS: The GitHub issue says that Linux must be used to convert
> //the key. Is this true? Is what I have here ^^ accurate and good enough for Win
> //users?

Author: ```suggestion
```
